### ISSUE TITLE
core.getObjectInfo("Sun")["altitude"] giving bad result?

### ISSUE THREAD

**Issue by NotSoRandomOne** [Stellarium/stellarium#778]
### Expected Behaviour
With following script, set Stellarium's time a minute or two after sunrise, then run script. I expect the script to stop when top of sun is just touching the horizon line. This is because your documentation for "core.getObjectInfo("Sun")["altitude"]" indicates it gives apparent altitude. Is it somehow overlooking refraction or something else?

### Actual Behaviour
Top of sun is below the horizon line, as shown in screen captures.

### Steps to reproduce
* Set to 'Zero horizon' with just horizon line showing.
* Make sure 'Atmosphere' is enabled.
* Go to a minute after sunrise
* Run the script. It will stop running when the sun is below the horizon, many seconds before true sunrise as given by other programs (or at least CyberSky, the one I checked against).

### System
* Stellarium version: 19.1.17157
* Operating system: Win10

![stellariumProblem](https://user-images.githubusercontent.com/12145990/65757917-af99d780-e0dd-11e9-816c-554128d1b0c6.jpg)

<pre>var sunDia = core.getObjectInfo("Sun")["size-dd"];
var desiredDepression = -sunDia/2;

var sunAlt = core.getObjectInfo("Sun")["altitude"];

while (sunAlt &gt; desiredDepression) {
   gotoTime("-1 second");
   sunAlt = core.getObjectInfo("Sun")["altitude"];
   }

function gotoTime(time) {
   core.setDate(time);
   waitForAltToChange();
   }
   
   
function waitForAltToChange() {
   var tempAlt = core.getObjectInfo("Sun")["altitude"];
   while (core.getObjectInfo("Sun")["altitude"] == tempAlt) core.wait(0.01);
   }</pre>

**Comment by gzotti** [comment]
Sun's vertical diameter is squashed by refraction. You would have to compute its apparent squashed size first. Else this is a nice query!

**Comment by NotSoRandomOne** [comment]
Have you ran across this calc before, and can point to it? Searching, I don't find it quickly, but will keep looking...  (Or is it already in Stellarium's code somewhere?)

**Comment by NotSoRandomOne** [comment]
Maybe there is an easier way to do what I am trying to do in Stellarium. If so, please let me know.

All I want is to see if the moon is visible on nights near newness. In the Planetary Visibility software, at http://www.alcyone.de/planetary_lunar_and_stellar_visibility.html, this is given as

visibility arc = 11.3° + 0.33 x distBetweenMoonSunCenters (where 11.3 and 0.33 can be adjusted for local conditions, which I don't really care about).

Stellarium makes the moon gradually appear as it separates from the sun, but does it have a method to check whether the moon meets a criteria like this (or another method) for first/last visibility?

I was planning on performing the above calc after finding sunset, but if it is already done, that would save the work.

Thanks,
David

**Comment by gzotti** [comment]
I can add a refraction-only scripting function later today (and in time for V0.19.2) for you to compute refraction effects to get the squashed solar diameter. 

distBetweenMoonSunCenters is elongation in our InfoMap

Note that some classical computation algorithms compute geometric sunrise/sunset without refraction. 

In general I really see you need this twilight/heliacal/lunar crescent visibility plugin (proposed name Eos [not related to any opto-mechanical device]) that I wanted to develop only AFTER fixing the last positional problems (we are now in general better than 1 arcmin, but lack things like aberration correction to get the seasons corrected). We could collaborate on that if you have the relevant data/models/papers already collected. I intend to play with (obviously) the 3 twilights (plus prob. some arbitrary depression), Arcus Visionis, Schoch, Schaefer, and as it seems now also crescent visibilities. E.g. there was a Danjon criterium for Lunar visibility, but I don't have the original sources. This could be (I think) developed over the next few months.

**Comment by NotSoRandomOne** [comment]
>> I can add a refraction-only scripting function later today (and in time for V0.19.2)...

:) Yay!!! Look forward to playing with it!

>> (we are now in general better than 1 arcmin, but lack things like aberration correction to get the seasons corrected)...

Searching for 'calculate sun's squashed size refraction' gave this PDF as the 5th result: https://pdfs.semanticscholar.org/027a/e4595f415f797c319d30b17225ef16cd04cb.pdf. It mentions aberration in it, so the book might be worth purchasing (as I believe the PDF probably violates copyright).

>> Schoch...

Swerdlow, et al, discuss Schoch's work here, with some values for deriving arcus visionis: http://www.alcyone.de/plsv/documentation/accuracy.html.

>> Schaefer...

Available here: http://adsabs.harvard.edu/full/1988QJRAS..29..511S

>> Danjon ...

Seems to just be a visibility labeling scheme without any calculation involved: https://eclipse.gsfc.nasa.gov/OH/OHres/Danjon.html.

Here's a note taken from my book I'm finishing up. In it, I illuminate the astronomy the biblical Samson story was based on 3,000 years ago.  The links may be helpful:

Endnote 7-14: 
C. Schoch, in “The ‘Arcus Visionis’ of the Planets in the Babylonian Observations,” gives 6.5° for evening last (Heliacal Setting), and 7.8° for morning first (Heliacal Rising) arcus visionis values for Sirius. (Monthly Notices of the Royal Astronomical Society, Vol. 84, 1924. Available online at http://adsabs.harvard.edu/full/1924MNRAS..84..731S.)
	Rita Gautschy, at http://www.gautschy.ch/~rita/archast/sirius/siriuseng.htm, gives Sirius’s arcus visionis first visibility in real terms for Egypt as being 2 to 3° above the horizon with the sun being depressed 6°.
	Using Gautschy’s criteria for the first illustration needed, a date of 12/28/1361 BC was found. Backing this up to the point where Sirius was on the horizon, the sun was then at about 3.4° below the horizon. I used a value of 4.0° in its place in order to quickly determine the first day of visibility. As a quick approximation, it should be good to within a few days for determining first visibility, even if Sirius is then at 2° elevation when it appears. Making the data tabular:
	
	Heliacal Rising	Heliacal Setting	Acronychal Rising	Cosmical Setting
Sirius	7.8°	6.5°	4.0°	4.0°

The reason for not directly using 3.4° is simply because I don’t trust the value. Given the earlier planetary arcus visionis table (note 4-15), it is hard to believe Sirius would appear at almost half of Jupiter’s Acronychal Rising and Cosmical Setting values. I may be wrong to be so untrusting, ...


>> ...that I wanted to develop only AFTER fixing the last positional problems...

I will probably be done with my work by the time you can get either of the projects well underway, so don't worry about rushing it.  I now have a way to fairly quickly give first and last lunar visibility dates. If you make a build with the new script feature enabled, I'll see if I can update the previous script to give true rise and set times, and visibility analysis.  Since the Samson myth wasn't even close to using arc-second precision, I don't need it in my work, and your outputs were close to CyberSky's.

If you, Alex, or anyone else on the Stellarium team are interested in my work, I am willing to send you PDFs when I've finished (in the next couple weeks), as long as you don't release them to the public. I can't afford to make them free to everyone at this point, as cracking our history open has left a lot of debt.  But you guys are welcome to it if you wish, for the time you've dedicated.

Have a great day!

**Comment by NotSoRandomOne** [comment]
PS, if you want to know more, circumstances conspired to make me do an interview before the book was finished. The Relevant or Irrelevant radio show: https://soundcloud.com/kalaradio/roi-show-324-full-program?in=kalaradio/sets/relevant-or-irrelevant. 28 mins long.

**Comment by gzotti** [comment]
Thanks for the pointers, good collection to start again. Yes, I would welcome if you PM me your book.

**Comment by NotSoRandomOne** [comment]
I don't see a way to PM in github. I assume your university email will work in a couple weeks when I'm done.

Have a good one!
David

**Comment by gzotti** [comment]
Yes, sure. :-) Best regards, Georg

**Comment by alex-w** [comment]
Georg, any news for this issue

**Comment by gzotti** [comment]
The commit 4af094f above added the requested refraction function (sorry, it was a bit late...) No news for the future Eos plugin mentioned.

**Comment by NotSoRandomOne** [comment]
For anyone who follows this up, a revised script that works, to save you time in 0.20.0. It is set up for sunrise, so for sunset you will need to change the time direction. Thank you, Georg, Alex, and anyone else involved!

<pre>var sunDia = core.getObjectInfo("Sun")["size-dd"];
var sunRadius = sunDia/2;
var sunAlt;
var refractedEdgeAlt;

updateVariables();

while (refractedAltEdge > 0) {
   gotoTime("-1 second");
   updateVariables();
   }


function updateVariables() {
   sunAlt = core.getObjectInfo("Sun")["altitude-geometric"];
   refractedAltEdge = core.refraction(sunAlt + sunRadius);
   }


function gotoTime(time) {
   core.setDate(time);
   waitForAltToChange();
   }
   
   
function waitForAltToChange() {
   var tempAlt = core.getObjectInfo("Sun")["altitude"];
   while (core.getObjectInfo("Sun")["altitude"] == tempAlt) core.wait(0.01);
   }</pre>
